scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,755 @@
+# mypy: disable-error-code="attr-defined"
+import numpy as np
+from scipy import special
+from scipy.optimize import OptimizeResult
+from scipy.optimize._zeros_py import (_scalar_optimization_initialize,
+                                      _scalar_optimization_loop,
+                                      _ECONVERGED, _ESIGNERR, _ECONVERR,  # noqa
+                                      _EVALUEERR, _ECALLBACK, _EINPROGRESS)  # noqa
+
+# todo:
+#  figure out warning situation
+#  address https://github.com/scipy/scipy/pull/18650#discussion_r1233032521
+#  without `minweight`, we are also suppressing infinities within the interval.
+#    Is that OK? If so, we can probably get rid of `status=3`.
+#  Add heuristic to stop when improvement is too slow / antithrashing
```

Comment:
Convergence is typically quite fast, so if we initially see convergence but it slows down before we meet tolerances, it's probably best to stop iterating. I'll wait for reviewer feedback before implementing such a heuristic, though.